fsharp-backend/src/LibExecution/Interpreter.fs

```diff
@@ -48,6 +53,327 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
+    }
+
+  let evalMatchNew id matchExpr cases =
+    let rec getMatchResult dv pattern builtUpTraces : MatchResult =
+      // Compare `dv` to `pattern`, and execute the rhs `expr` of any
+      // matches. Tracks whether a branch has already been executed and
+      // will exceute later matches in preview mode.  Ensures all patterns
+      // and branches are properly traced.  Recurse on partial matches
+      // (constructors); builtUpTraces is the set of traces that have been
+      // built up by recursing: they can only be matched when the pattern
+      // is ready to match.
+      // TODO: revise the above; maybe move it or break it down
+      match pattern with
+      | PInteger (pid, i) ->
+        let v = DInt i
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PBool (pid, bool) ->
+        let v = DBool bool
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PCharacter (pid, c) ->
+        let v = DChar(c)
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PString (pid, str) ->
+        let v = DStr(str)
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PFloat (pid, v) ->
+        let v = DFloat v
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PNull (pid) ->
+        let v = DNull
+
+        if v = dv then
+          Matched((pid, v) :: builtUpTraces, [])
+        else
+          Unmatched(v, builtUpTraces)
+      | PVariable (pid, v) ->
+        // only matches allowed values
+        if Dval.isFake dv then
+          Unmatched(dv, builtUpTraces)
+        else
+          Matched((pid, dv) :: builtUpTraces, [ (v, dv) ])
+      | PBlank (pid) ->
+        // never matches
+        Unmatched(incomplete pid, builtUpTraces)
+      | PConstructor (pid, name, args) ->
+        match (name, args, dv) with
+        | "Just", [ p ], DOption (Some v)
+        | "Ok", [ p ], DResult (Ok v)
+        | "Error", [ p ], DResult (Error v) ->
+          getMatchResult v p ((pid, dv) :: builtUpTraces)
+        | "Nothing", [], DOption None -> Matched((pid, dv) :: builtUpTraces, [])
+        | "Nothing", [], _ -> Unmatched(DOption None, builtUpTraces)
+        | _ ->
+          let error =
+            if List.contains name [ "Just"; "Ok"; "Error"; "Nothing" ] then
+              incomplete pid
+            else
+              DError(sourceID pid, $"Invalid constructor: {name}")
+
+          // Trace each argument too. TODO: recurse
+          // List.iter
+          //   (fun pat ->
+          //     let id = Pattern.toID pat
+          //     state.tracing.traceDval false id (incomplete id))
+          //   args
```

Comment:
So far, I've been ignoring this now-commented out trace logic. I'm curious if we need it, and if so, how to fit it into this model. (this may be clear to me later)

Reply:
To determine if we need it, you could play around with the editor and see what patterns have analysis results available to them. If everything looks right, then that logic probably isn't needed